mitodl/mit-open
# Learningpaths UI

# What are the relevant tickets?
#61 

# Description (What does it do?)
This PR converts the infinite-corridor "stafflists" frontend UI to use the new `/learningpaths` API.

Some notes:
- UI Change: All learningpaths are sortable. (No distinction between "paths" and "lists". If authors don't want to sort the list, they don't have to.)
- I deleted the old stafflists / userlist code, despite the fact that this PR does not convert userlists to the new API. Since we're no longer treating userlists as learning resources, the frontends of learningpaths / userlists may diverge more. Also, since the old code is easily available in the open-discussions repo, I don't see much advantage to keeping it around here with deprecated course-catalog APIs.
- Learningpaths do not yet have an associated image

# Screenshots (if appropriate):

<img width="700" alt="AddToListDialog" src="https://github.com/mitodl/mit-open/assets/9010790/1170f4f2-ea15-481e-b908-4c7855100e1c">
<img width="700" alt="details_page" src="https://github.com/mitodl/mit-open/assets/9010790/3a1e3304-e4ea-4103-a967-d74e0d78bda9">
<img width="700" alt="edit_list" src="https://github.com/mitodl/mit-open/assets/9010790/1f2120f1-f157-4d34-843e-479e9be50f9a">
<img width="700" alt="listing" src="https://github.com/mitodl/mit-open/assets/9010790/b972f6dc-4ad9-47b7-8840-fc5d626b1d45">
<img width="700" alt="sorting" src="https://github.com/mitodl/mit-open/assets/9010790/259d3a53-5ada-4539-8362-98733d50e051">


# How can this be tested?
Viewing the site as an admin (or anyone with `learning_path_editor` permissions):

1. Create a few learning paths at http://localhost:8063/learningpaths
    - Try various options: public, private, etc.
    - Try saving without required fields. You should get an error.
2. Adding/removing resources to/from your learning paths:
    - The only courses you can add/remove via UI are the ones on homepage; use the `+` button on resource cards.
    - You can add more resources via POST requests at http://localhost:8063/api/v1/learningpaths/PATH_ID_HERE/resources/
    - Once added, you can remove via `+` button dialog, too, either on the homepage or on the details page (see below).
3. View http://localhost:8063/learningpaths/PATH_ID_HERE/resources/ to view the resources added to your learning paths.
    - you can add/remove via `+` button
    - Try sorting the list items ("Reorder")

**Notes:** None of the listings above are paginated as of now

**As an anonymous user:** Or at least someone without `learning_path_editor`:

1. Check that  http://localhost:8063/learningpaths/ only lists public paths and has no "Create" button. 
2. Check that the "details page" (http://localhost:8063/learningpaths/PATH_ID_HERE/resources/) for public paths are viewable, but not viewable for private paths.
    - details page should not have Reorder / Edit buttons
